Fetch shellcheck from vscode-shellcheck for M1 support

gauthamnair opened the pull request:
Fixes #18589 

https://github.com/pantsbuild/pants/issues/18589#issuecomment-1737616128
> Shellcheck is available from VSCode: https://github.com/vscode-shellcheck/shellcheck-binaries/releases
> 
> and this is the recommendation of shellcheck itself:
> 
> https://github.com/koalaman/shellcheck/blob/90d3172dfec30a7569f95b32479ae97af73b8b2e/README.md?plain=1#L236-L237
> 
> > There are currently no official binaries for Apple Silicon, but third party builds are available via [ShellCheck for Visual Studio Code](https://github.com/vscode-shellcheck/shellcheck-binaries/releases).
> 

## Testing

This has fixed shellcheck in my local env without the need for rosetta.

gauthamnair commented on src/python/pants/backend/shell/lint/shellcheck/subsystem.py:
```diff
@@ -20,18 +20,18 @@ class Shellcheck(TemplatedExternalTool):
+        "v0.8.0|macos_x86_64|4e93a76ee116b2f08c88e25011830280ad0d61615d8346364a4ea564b29be3f0|6310442",
+        "v0.8.0|linux_arm64 |8f4810485425636eadce2ec23441fd29d5b1b58d239ffda0a5faf8dd499026f5|4884430",
+        "v0.8.0|linux_x86_64|01d181787ffe63ebb0a2293f63bdc8455c5c30d3a6636320664bfa278424638f|2082242",
     ]
```
The shas and sizes have changed. The structure of the packaged dir is different. It is now a tar.gz that contains only the `shellcheck` binary within.

huonw replied:
For the platforms that are shared, is the shellcheck binary the same? Or does its hash change too?

gauthamnair replied:
I just now checked, and it turns out that the sha of the binary contained inside the archive _is_ exactly the same for the platforms that are shared. The difference in sha and size of the archive must be because of the difference in compression and/or directory substructure.

slightly formatted results, showing the sha | bytecount of the archive and of the extracted executable:
```
## macos_arm64
https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/v0.8.0/shellcheck-v0.8.0.darwin.aarch64.tar.gz
36dffd536b801c8bab2e9fa468163037e0c7f7e0a05298e5ad6299b4dde67e31 | 14525367
shellcheck
02dd463e665d0c20a243c600f0d93dd3fd90d20e02355b06660cc18c9da884f5 | 74608983

## macos_x86_64
https://github.com/koalaman/shellcheck/releases/download/v0.8.0/shellcheck-v0.8.0.darwin.x86_64.tar.xz
e065d4afb2620cc8c1d420a9b3e6243c84ff1a693c1ff0e38f279c8f31e86634 | 4049756
shellcheck-v0.8.0/shellcheck
5e0dbf0166a30194f48ec74e453eefaa0807f8150b728281b70a9b91ec2b38da | 32603580

https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/v0.8.0/shellcheck-v0.8.0.darwin.x86_64.tar.gz
4e93a76ee116b2f08c88e25011830280ad0d61615d8346364a4ea564b29be3f0 | 6310442
shellcheck
5e0dbf0166a30194f48ec74e453eefaa0807f8150b728281b70a9b91ec2b38da | 32603580

## linux_arm64
https://github.com/koalaman/shellcheck/releases/download/v0.8.0/shellcheck-v0.8.0.linux.aarch64.tar.xz
9f47bbff5624babfa712eb9d64ece14c6c46327122d0c54983f627ae3a30a4ac | 2996468
shellcheck-v0.8.0/shellcheck
f0dd2c031db171e6e95fd7b53c6ff32ff96c94e1fb40c5f28a23b50f72f57c89 | 20628560

https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/v0.8.0/shellcheck-v0.8.0.linux.aarch64.tar.gz
8f4810485425636eadce2ec23441fd29d5b1b58d239ffda0a5faf8dd499026f5 | 4884430
shellcheck
f0dd2c031db171e6e95fd7b53c6ff32ff96c94e1fb40c5f28a23b50f72f57c89 | 20628560

## linux_x86_64
https://github.com/koalaman/shellcheck/releases/download/v0.8.0/shellcheck-v0.8.0.linux.x86_64.tar.xz
ab6ee1b178f014d1b86d1e24da20d1139656c8b0ed34d2867fbb834dad02bf0a | 1403852
shellcheck-v0.8.0/shellcheck
f4bce23c11c3919c1b20bcb0f206f6b44c44e26f2bc95f8aa708716095fa0651 | 6753016

https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/v0.8.0/shellcheck-v0.8.0.linux.x86_64.tar.gz
01d181787ffe63ebb0a2293f63bdc8455c5c30d3a6636320664bfa278424638f | 2082242
shellcheck
f4bce23c11c3919c1b20bcb0f206f6b44c44e26f2bc95f8aa708716095fa0651 | 6753016
```


Script used to produce the data for the above:
```sh
#!/bin/bash

set -e

OUTPUT_BASE_DIR="gsecrets"

VSCODE_BASE_DIR="${OUTPUT_BASE_DIR}/shellcheck-tests/vscode"
KOALAMAN_BASE_DIR="${OUTPUT_BASE_DIR}/shellcheck-tests/koalaman"

mkdir -p "${VSCODE_BASE_DIR}"
mkdir -p "${KOALAMAN_BASE_DIR}"

download_shellcheck_vscode() {
  version=$1
  platform=$2
  filename="shellcheck-${version}.${platform}.tar.gz"
  URL="https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/${version}/${filename}"
  echo "$URL"
  curl -s -L -o "${VSCODE_BASE_DIR}/${filename}" "${URL}"
  shasum -a 256 "${VSCODE_BASE_DIR}/${filename}"
  wc -c "${VSCODE_BASE_DIR}/${filename}"

  mkdir -p "${VSCODE_BASE_DIR}/${version}-${platform}"
  tar -xzf "${VSCODE_BASE_DIR}/${filename}" -C "${VSCODE_BASE_DIR}/${version}-${platform}" "shellcheck"
  shasum -a 256 "${VSCODE_BASE_DIR}/${version}-${platform}/shellcheck"
  wc -c "${VSCODE_BASE_DIR}/${version}-${platform}/shellcheck"
}

download_shellcheck_koalaman() {
  version=$1
  platform=$2
  filename="shellcheck-${version}.${platform}.tar.xz"
  URL="https://github.com/koalaman/shellcheck/releases/download/${version}/${filename}"
  echo "$URL"
  curl -s -L -o "${KOALAMAN_BASE_DIR}/${filename}" "${URL}"
  shasum -a 256 "${KOALAMAN_BASE_DIR}/${filename}"
  wc -c "${KOALAMAN_BASE_DIR}/${filename}"

  mkdir -p "${KOALAMAN_BASE_DIR}/${version}-${platform}"
  tar -xzf "${KOALAMAN_BASE_DIR}/${filename}" -C "${KOALAMAN_BASE_DIR}/${version}-${platform}" "shellcheck-${version}/shellcheck"
  shasum -a 256 "${KOALAMAN_BASE_DIR}/${version}-${platform}/shellcheck-${version}/shellcheck"
  wc -c "${KOALAMAN_BASE_DIR}/${version}-${platform}/shellcheck-${version}/shellcheck"
}

echo 'macos_arm64'
download_shellcheck_vscode v0.8.0 darwin.aarch64
echo ''

echo 'macos_x86_64'
download_shellcheck_koalaman v0.8.0 darwin.x86_64
download_shellcheck_vscode v0.8.0 darwin.x86_64

echo 'linux_arm64'
download_shellcheck_koalaman v0.8.0 linux.aarch64
download_shellcheck_vscode v0.8.0 linux.aarch64

echo 'linux_x86_64'
download_shellcheck_koalaman v0.8.0 linux.x86_64
download_shellcheck_vscode v0.8.0 linux.x86_64
```
I also spot-checked the darwin.86 one with plain downloading from the browser and manual running of commands in case I flubbed something up with the script above.

huonw replied:
Nice! Thanks for verifying it. It's good that they match exactly. 👍

gauthamnair commented on src/python/pants/backend/shell/lint/shellcheck/subsystem.py:
```diff
@@ -52,7 +52,7 @@ class Shellcheck(TemplatedExternalTool):
 
     def generate_exe(self, _: Platform) -> str:
-        return f"./shellcheck-{self.version}/shellcheck"
+        return f"./shellcheck"
```
The `tar.xz` from shellcheck looked like:
```
~/Downloads $ tar -tzf shellcheck-v0.8.0.darwin.x86_64.tar.xz
shellcheck-v0.8.0/LICENSE.txt
shellcheck-v0.8.0/README.txt
shellcheck-v0.8.0/shellcheck
```
but the ones from vscode are just:
```
~/Downloads $ tar -tzf shellcheck-v0.8.0.darwin.aarch64.tar.gz
shellcheck
```

huonw commented on src/python/pants/backend/shell/lint/shellcheck/subsystem.py:
```diff
@@ -20,18 +20,18 @@ class Shellcheck(TemplatedExternalTool):
     default_url_template = (
-        "https://github.com/koalaman/shellcheck/releases/download/{version}/shellcheck-"
-        "{version}.{platform}.tar.xz"
+        "https://github.com/vscode-shellcheck/shellcheck-binaries/releases/download/{version}/shellcheck-"
```
Maybe we can as a comment about why we're using an "unofficial" source?

huonw reviewed (commented):
Thanks for this! It looks like flake8 is asking for some adjustment too.

gauthamnair commented:
@huonw I incorporated your suggestions and did the check of the extracted executables. The lint problem is now fixed but I have two test failures:

```
✕ src/python/pants/backend/python/goals/coverage_py_integration_test.py:tests failed in 540.07s.
✕ src/python/pants/init/load_backends_integration_test.py:tests failed in 900.03s.**
```
They are both test time out issues:
```
        "coverage_py_integration_test.py": {
            "tags": ["platform_specific_behavior"],
            "timeout": 540,
        },
```
and 
```
"load_backends_integration_test.py": {"timeout": 900},
```
Locally the coverage_py integration test runs fine for me. (albeit slow at 180s even though it is only testing against python3.9.

Should we increase the timeout on the coverage integration test? Kind of puzzling to me why they are so slow in the first place though.

huonw reviewed (approved):
Thanks for this! It seems to work well on my Apple Silicon Mac.

It's a bit unfortunate shellcheck doesn't release Mac aarch64 binaries (https://github.com/koalaman/shellcheck/issues/2714), but the hashes matching (thanks for verifying!) and it being linked from the main repo seems good enough. ✅

huonw commented:
(You're right that the timeouts could be increased as a separate PR, to at least reduce the instances of spurious flaky/inconsistent failures on unrelated PRs. In this case, the retry I just did passed fine.)

purajit commented:
I understand why this was needed, but it seems very weird to me that we're using a _third-party_ build that was only recommended for Apple Silicon for _all_ platforms, especially one that bills itself as "ShellCheck for Visual Studio Code". It makes sense for individual projects to choose that, but changing that at a base-build-framework level doesn't seem right to me. It is very clearly not recommended to use this build for non-Apple Silicon platforms, even according to the `shellcheck` README itself.

I would prefer overriding the URL in this one case; I can PR that if people are in support of that.

Happened to catch this since we do some base pants image builds for speed, and suddenly got a bunch of non-matching-digest errors when upgrading pants.
